src/qibo/models/dbi/utils.py

```diff
@@ -162,3 +101,54 @@ def cs_angle_sgn(dbi_object, d):
         )
     )
     return np.sign(norm)
+
+
+def decompose_into_Pauli_basis(h_matrix: np.array, pauli_operators: list):
+    """finds the decomposition of hamiltonian `h_matrix` into Pauli-Z operators"""
+    nqubits = int(np.log2(h_matrix.shape[0]))
+
+    decomposition = []
+    for Z_i in pauli_operators:
+        expect = np.trace(h_matrix @ Z_i) / 2**nqubits
+        decomposition.append(expect)
+    return decomposition
+
+
+def generate_pauli_index(nqubits, order):
+    if order == 1:
+        return list(range(nqubits))
+    elif order > 1:
+        indices = list(range(nqubits))
+        return indices + [
+            comb for i in range(2, order + 1) for comb in combinations(indices, i)
+        ]
+    else:
+        raise ValueError("Order must be a positive integer")
+
+
+def generate_pauli_operator_dict(
+    nqubits: int, parameterization_order: int = 1, symbols_pauli=symbols.Z
+):
+    pauli_index = generate_pauli_index(nqubits, order=parameterization_order)
+    pauli_operators = [
+        generate_Pauli_operators(nqubits, symbols_pauli, index) for index in pauli_index
+    ]
+    return {index: operator for index, operator in zip(pauli_index, pauli_operators)}
+
+
+def diagonal_min_max(matrix: np.array):
+    L = int(np.log2(matrix.shape[0]))
+    D = np.linspace(np.min(np.diag(matrix)), np.max(np.diag(matrix)), 2**L)
+    D = np.diag(D)
+    return D
+
+
+def generate_Pauli_operators(nqubits, symbols_pauli, positions):
+    # generate matrix of an nqubit-pauli operator with `symbols_pauli` at `positions`
+    if isinstance(positions, int):
+        return SymbolicHamiltonian(
+            symbols_pauli(positions), nqubits=nqubits
+        ).dense.matrix
+    else:
+        terms = [symbols_pauli(pos) for pos in positions]
+        return SymbolicHamiltonian(math.prod(terms), nqubits=nqubits).dense.matrix
```

Comment:
Please provide docstrings for those functions

Reply:
@Sam-XiaoyueLi when you have the time, please take a look at the docstrings I made to see if they accurately describe the functions :)